### ISSUE TITLE
What about semantic versioning?

### ISSUE THREAD

**Issue by jbuberel** [matttproud/gochecklist#1]
Have you thought about adding a checklist for establishing a set of semantic versioning rules?

Having spent too many years in the Java world, one nice thing you could count on is that nearly all Java libraries used the MAJOR.MINOR.FIX release notation. This is something that I believe could benefit the Go community.

While I understand that the Go community's solution to this is "vendor the version you depend on", I still believe there is value in naming and labeling specific releases. 

There is a generally accepted rule within Go that once you reach a 1.0.0 version, you are making a guarantee of API compatibility. There is also a suggestion that if you want to change an API after 1.0, you need to do that using a new import path, as in:

```
import "github.com/myuser/mylibrary/v1"
import "github.com/myuser/mylibrary/v2"
```

Do you think this would be useful for Go? Too much complexity?

**Comment by matttproud** [comment]
Hey @jbuberel, thank you for raising the question.  I had been mulling over it myself but hadn't reached a definitive conclusion.  Here is what I propose doing:

Under the [Good Citizen](https://github.com/matttproud/gochecklist/blob/master/README.md#good-citizen) guidance section, we have a blurb that speaks about _API Stability and Maturity Grants_.  That topic seems like a perfect place to discuss this.  Would you feel interested in fleshing out a draft `publication/api_stability_and_maturity.md` file that discusses this more deeply?

A couple of messages/values that speak out to me in terms of how this document should be worded:
1. There are multiple disciplines that can be chosen.  The author should pick one and stick to it consistently, trying to prefer well-known, preexisting conventions—like semver—as a default.
2. The author should mention something about the overall maturity of the project, what is likely to change, when, and possibly how.  Is it in use for production, a prototype, legacy, etc?
3. I would refrain from suggesting a particular methodology for exposing the versioning to the project's customers (e.g., version control or filesystem path literals that convey a frozen point in time).  The Go community has yet to come to full concensus itself on how to do this, if at all.  Fads explode and fade as quickly as fireworks in the night.  What does seem worth mentioning is version control release tagging, however, since it is agnostic to most of the common vendoring schemes people use.  (Vendoring and stability grants are closely related but also sufficiently orthogonal.)

Does this seem fair?

I had been thinking about writing an article about implementing one's own `go fix` tool to rewrite APIs for customers en masse, but only big projects that are in transition would care about this …

**Comment by jbuberel** [comment]
@matttproud - Happy to help. Let me take a pass at the first draft.

What are your thoughts on the following rule, which Rob is fond of quoting:

> Once you reach a stable API, any breaking changes to that API should mean the import path changes.

You also have Eric Fowler who insists that you should never include API version information in your API request path. He suggests that versioning should be done at the level of the hostname (api.github.com vs api2.github.com). I'm not sure I would go that far though.

**Comment by matttproud** [comment]
Sounds great.  Can't wait to see what you come up with!

**Comment by jbuberel** [comment]
Take a look at this first draft, and let me know what you think:

https://github.com/jbuberel/gochecklist/blob/master/publication/api_stability_and_maturity.md

**Comment by matttproud** [comment]
Cool!  Looks great overall!  Want to take care of these items and then open a pull request?

## Major Remarks
1. In the `Project Maturity and Stability` section, I might include a blurb about doing a favor to users to mention what could or is likely to change—provided the remarking of such things does not entail significant overhead.  APIs development is an iterative process, so sometimes reflection in the thereafter produces profound wisdom about the _next time_.
2. In the `Semantic Versioning` section, I might make a footnote that another option is to _expect users to take appropriate steps, whatever they deem it for themselves to vendor their dependencies_ and just break the API for the base bath, especially moving from `v1` to `v2` is a rather self-explanatory exercise.  The core requirement, though, is that source control revision IDs and tags are used and meaningful.  Historically this has been my preference _when the migration is abundantly straightforward_.  When I needed to change a project repository's internal layout, I have created a [moved.go](https://github.com/matttproud/golang_protobuf_extensions/blob/master/ext/moved.go) that is [similar to what upstream Go has periodically done](https://code.google.com/p/go/source/browse/present/moved.go?spec=svn.talks.1a303e2eaf127ae0a75d5998b44f719d8f3e6176&repo=talks&r=1a303e2eaf127ae0a75d5998b44f719d8f3e6176) since this is one of those trivial steps.
3. This should be linked from the main `README.md`.

## Nits
1. Typo of `langugae`.
2. Mind attempting to break lines, where doing so does not detract from readability, to approximately 80 characters?
3. Mind using double spaces after period?  Markdown renderers will do the correct thing, but many folks view the content in fixed-width editors.  This will keep it consistent with the rest of the corpus?

**Comment by matttproud** [comment]
I have gone ahead and merged #5, so I think we can mark this as fixed for now!
